src/Units/AWBWWeapons.java

```diff
@@ -13,7 +13,7 @@
     CARRIERMISSILES, BATTLESHIPCANNON, CRUISERTORPEDOES, CRUISERMGUN, SUBTORPEDOES
   };
   
-  protected final static boolean USE_AMMO = true;
+  protected final static boolean USE_AMMO = false;
```

Comment:
Maybe change this to `INFINITE_AMMO` so it matches the function it's passed into? Probably the wrong value was used because they are named to disagree.

Reply:
If I called it `INFINITE_AMMO`, I'd have to put exclamation points in front of all of its current usages.
I'm just glad I decided to test running out of ammo, heh.

Reply:
... or you could change the default value back, and enjoy your newfound sense of alignedness.

Reply:
I named it `USE_AMMO` to be idiomatic in the `WeaponModel` constructors. I tell it to `USE_AMMO`, or I don't and it doesn't. Matching the declaration of the constructor wasn't really a priority.
I'm not sure what you mean by that last comment.

Reply:
> I named it USE_AMMO to be idiomatic in the WeaponModel constructor calls

What? The variable name in `WeaponModel`'s constructor is called `infiniteAmmo`. The member variable is named `hasInfiniteAmmo`. Renaming this one to match would (aligning them) would make it clear exactly what we are doing.
Right now, `USE_AMMO==false` corresponds to `infiniteAmmo==true`, which adds an extra layer of confusion re-aligning the concepts in your head. Making it matchy-matchy makes it easier.
As for being idiomatic, I'd argue renaming this to match would be *more* idiomatic, for two reasons:
1) A new dev who looked at this wouldn't necessarily connect the variable name `USE_AMMO` to the concept of "having a limited quantity", whereas a variable named `HAS_INFINITE_AMMO` is exceedingly clear.
2) `USE_AMMO` is set to false. Anybody who sees a Boolean value set to false is going to assume that means the thing we pass it to is being told not to "use ammo". If they manage to connect it to the concept of "having infinite ammo", then the weapon is clearly being told to not use ammo, meaning it should have infinite ammo; the opposite is true.

There's a reason naming things is one of the two Hard Problems in computer science.

Reply:
If I call it `HAS_INFINITE_AMMO`, my pea-brain will look at the `WeaponModel` constructors and panic because it "says right there" that the tank cannon gets infinite ammo.
I suppose I could meet you in the middle with `DISABLE_INFINITE_AMMO`?

> Right now, `USE_AMMO`==false corresponds to `infiniteAmmo`==true, which adds an extra layer of confusion re-aligning the concepts in your head.

That's not true at all.
`USE_AMMO` *being present* corresponds with `infiniteAmmo`==true.
`USE_AMMO` being *not* there corresponds with `infiniteAmmo`==false.
Honestly, I don't care all that much about aligning concepts in this case. I just want to not have to look at it :stuck_out_tongue:

> A new dev who looked at this wouldn't necessarily connect the variable name `USE_AMMO` to the concept of "having a limited quantity",

Perhaps a better term would be `REQUIRES_AMMO`, then?

> USE_AMMO is set to false. Anybody who sees a Boolean value set to false is going to assume that means the thing we pass it to is being told not to "use ammo". If they manage to connect it to the concept of "having infinite ammo", then the weapon is clearly being told to not use ammo, meaning it should have infinite ammo; the opposite is true.

That's almost exactly the opposite of how my brain works.
If I see a weapon constructor, and it says `USE_AMMO`, brain says "yay good, we use ammo".
If I decide to look deeper and see that that boolean happens to be false, I'm going to look at where it's used and quickly arrive at the right conclusion.

Reply:
> That's not true at all.

Hahaha, except it is; you just reiterated what I said. observe:
> USE_AMMO being present corresponds with infiniteAmmo==true.

If `USE_AMMO` is *present*, it is false, which corresponds with having infinite ammo. Thanks for helping to make my point about the names being confusing.

> I suppose I could meet you in the middle with DISABLE_INFINITE_AMMO

Ah, I hadn't looked closely enough to notice you only pass `USE_AMMO` to weapons with limited ammo. Yes, I can meet you in the middle there. Or `REQUIRES_AMMO` or `LIMITED_AMMO` or `EATS_BULLETS` or whatever (except not that last one please).

> That's almost exactly the opposite of how my brain works.

Aaaaand, this is why naming things is hard. You have to be anticipate potential misconceptions and methodically destroy them, leaving a single unambiguous interpretation. If you only write code for yourself, nobody else will be able to read it.

Reply:
> You have to be anticipate potential misconceptions and methodically destroy them

Actually, I should clarify that this is *not* the only approach. Another approach I've encountered is to use perfectly ambiguous names that can grant no preconceptions. E.g. if you change the variable to be named `SILLY_KANGAROO`, then nobody would be able to guess at what it meant, and they'd have to look at how it was used.
A side effect of this style of naming things is that conversations about the code can get weird. "OK, we've got a new unit, and it's got a new type of weapon." "Is it a silly kangaroo weapon, or a normal one?" etc.

So uh, yeah.

Reply:
> If `USE_AMMO` is _present_, it is false, which corresponds with having infinite ammo. Thanks for helping to make my point about the names being confusing.

It's only confusing if you look exactly one level deep. If you look zero or two levels deep, the meaning is abundantly clear.
0: "the gun says it uses ammo"
1: "the gun says uses ammo=false, so it must have infinite"
2: "the constructor parameter is called infinite ammo, so it's actually *disabling* infinite ammo"

> Ah, I hadn't looked closely enough to notice you only pass `USE_AMMO` to weapons with limited ammo. Yes, I can meet you in the middle there. Or `REQUIRES_AMMO` or `LIMITED_AMMO` or `EATS_BULLETS` or whatever (except not that last one please).

What would be the point of defining a value that's constant for *all* weapons if I was going to pass it to every single weapon's constructor? At that point, I'd just remove it from the constructor and make it an implied parameter.
...you should know by now to not give me ideas like that.

> Aaaaand, this is why naming things is hard. You have to be anticipate potential misconceptions and methodically destroy them, leaving a single unambiguous interpretation. If you only write code for yourself, nobody else will be able to read it.

I can't anticipate every misconception. Even if I could, it'd be a lot harder than actually getting the code working anyhow.
I've actually revised the name of that variable more than once already in pursuit of clarity.

> Actually, I should clarify that this is _not_ the only approach. Another approach I've encountered is to use perfectly ambiguous names that can grant no preconceptions. E.g. if you change the variable to be named `SILLY_KANGAROO`, then nobody would be able to guess at what it meant, and they'd have to look at how it was used.

You're just piling more evidence on the docket for why we should use `EATS_BULLETS`.